RFC: Add embassy net support

oleid opened the pull request:
This is a little something I cooked up for little embassy based esp32 project.
Using these changes I was able to get times from a timeserver.

embassy-net switched to core::net addresses recently, however, smoltcp (which is used under the hood) didn't switch completely, hence some type conversion is still needed.

What do you think? Would you consider supporting embassy-net in this crate?

vpetrigo commented:
Hey @oleid!
Thank you for the PR. I will review the approach you suggested and get back with a feedback. Since every `net` implementations (e.g. in `tokio`, in `embassy`) introduce a crate specific types almost everytime, maybe it would be better to have a `core::net::SocketAddr` requirement for the public interface and a user is free to implement the part that is required for theirs application.

oleid commented:
> Since every net implementations (e.g. in tokio, in embassy) introduce a crate specific types almost everytime, maybe it would be better to have a core::net::SocketAddr requirement for the public interface and a user is free to implement the part that is required for theirs application.

That makes a lot of sense, actually.

I see failing tests... I can take a look to see what's wrong.

dpeckett commented:
Following on to this, I've refactored the NtpUdpSocket trait to included the suggested changes from @vpetrigo, see #35

As I get a bit further along with this, perhaps we can add an embassy-net example, showing how to implement the NtpUdpSocket trait.

One of the limitations of the no_std ToSocketAddrs implementations from @oleid  is that DNS resolution is left to the user. I think this probably makes sense as there's no good way of doing this in a generic / runtime independent fashion.

oleid commented:
> Following on to this, I've refactored the NtpUdpSocket trait to included the suggested changes from @vpetrigo, see #35

Great, I rebased to your branch.
 
> As I get a bit further along with this, perhaps we can add an embassy-net example, showing how to implement the NtpUdpSocket trait.

I thought about this a little. It is possible to build embassy-net as a linux application, maybe this would make a sane example.
 
> One of the limitations of the no_std ToSocketAddrs implementations from @oleid is that DNS resolution is left to the user. I think this probably makes sense as there's no good way of doing this in a generic / runtime independent fashion.

That ToSocketAddrs implementation was stolen from no_std-net, by the way. But yes, I agree that should be left to the user. Maybe one could add a helper function to the net module that uses ToSocketAddrs. That would make the examples a little simpler. Otherwise they would need to contain name resolution like here (I started refactoring the examples a little):

```diff
--- a/examples/timesync/src/main.rs
+++ b/examples/timesync/src/main.rs
@@ -19,7 +19,8 @@
 //!
 //! Example provides a basic implementation of [`NtpTimestampGenerator`] and [`NtpUdpSocket`]
 //! required for the [`sntpc`] library
-use std::net::UdpSocket;
+use anyhow::Context;
+use std::net::{ToSocketAddrs, UdpSocket};
 use std::str::FromStr;
 use std::time::Duration;
 
@@ -27,7 +28,7 @@ use clap::{crate_version, App, Arg};
 
 const GOOGLE_NTP_ADDR: &str = "time.google.com";
 
-fn main() {
+fn main() -> anyhow::Result<()> {
     let app = App::new("timesync")
         .version(crate_version!())
         .arg(
@@ -56,25 +57,23 @@ fn main() {
     }
 
     let ntp_server = app.value_of("server").unwrap();
-    let ntp_port = u32::from_str(app.value_of("port").unwrap());
-
-    let ntp_port = match ntp_port {
-        Ok(ntp_port) => ntp_port,
-        Err(err) => {
-            eprintln!("Unable to convert NTP server port value: {err}");
-            return;
-        }
-    };
+    let ntp_port = u32::from_str(app.value_of("port").unwrap())
+        .context("While parsing NTP server port")?;
 
     let ntp_addr = format!("{ntp_server}:{ntp_port}");
+    let addr = ntp_addr
+        .to_socket_addrs()?
+        .next()
+        .context("While resolving NTP server address")?;
 
     let socket =
         UdpSocket::bind("0.0.0.0:0").expect("Unable to create UDP socket");
     socket
         .set_read_timeout(Some(Duration::from_secs(2)))
         .expect("Unable to set UDP socket read timeout");
-    let time = sntpc::simple_get_time(ntp_addr.as_str(), &socket)
+    let time = sntpc::simple_get_time(addr, &socket)
         .unwrap_or_else(|_| panic!("Unable to receive time from: {ntp_addr}"));
 
     sntpc::utils::update_system_time(time.sec(), time.sec_fraction());
+    Ok(())
 }
```

vpetrigo commented:
Looks good to me! Thank you for elaborating on that.
